Add ASP.NET Core release notes for .NET 10 Preview 5 (#9925)
dotnet/core: release-notes/10.0/preview/preview5/aspnetcore.md

# ASP.NET Core in .NET 10 Preview 5 - Release Notes

Here's a summary of what's new in ASP.NET Core in this preview release:

> - [Configure Custom Security Descriptors for HTTP.sys Request Queues](#configure-custom-security-descriptors-for-httpsys-request-queues)

Reviewer: Use sentence casing for all headings.

> - [Validation in minimal APIs](#validation-in-minimal-apis)

Reviewer: Let's change this section to "Validation resolver APIs marked as experimental".

// Create a discretionary access control list (DACL)
var dacl = new DiscretionaryAcl(isContainer: false, isDS: false, capacity: 2);
dacl.AddAccess(
    AccessControlType.Allow,
    new SecurityIdentifier(WellKnownSidType.BuiltinUsersSid, null),
    -1,
    InheritanceFlags.None,
    PropagationFlags.None
);
dacl.AddAccess(
    AccessControlType.Deny,
    new SecurityIdentifier(WellKnownSidType.BuiltinGuestsSid, null),
    -1,
    InheritanceFlags.None,
    PropagationFlags.None
);

// Assign the DACL to the security descriptor
securityDescriptor.DiscretionaryAcl = dacl;

// Configure HTTP.sys options
var builder = WebApplication.CreateBuilder();
builder.WebHost.UseHttpSys(options =>
{
    options.RequestQueueSecurityDescriptor = securityDescriptor;
});
```

### Additional notes

- The `RequestQueueSecurityDescriptor` applies only when creating a new request queue.
- This property does not affect existing request queues.
- See the official documentation for more information about Windows security descriptors and their usage.

## Validation resolver APIs marked as experimental

> A number of small improvements and fixes have been made to the validation generator for Minimal APIs that was introduced in preview 4.

Reviewer: ```suggestion
```

# ASP.NET Core in .NET 10 Preview 5 - Release Notes

Here's a summary of what's new in ASP.NET Core in this preview release:

- [Configure custom security descriptors for HTTP.sys request queues](#configure-custom-security-descriptors-for-httpsys-request-queues)
- [Validation resolver APIs marked as experimental](#validation-resolver-apis-marked-as-experimental)
- [Support for generating OpenAPI 3.1](#support-for-generating-openapi-31)
- [OpenAPI metadata from XML doc comments](#openapi-metadata-from-xml-doc-comments)
- [Add a Not Found page using the Blazor Router](#add-a-not-found-page-using-the-blazor-router)
- [Blazor metrics and tracing](#blazor-metrics-and-tracing)

ASP.NET Core updates in .NET 10:

- [What's new in ASP.NET Core in .NET 10](https://learn.microsoft.com/aspnet/core/release-notes/aspnetcore-10.0) documentation.
- [Breaking changes](https://docs.microsoft.com/dotnet/core/compatibility/10.0#aspnet-core)
- [Roadmap](https://github.com/dotnet/aspnetcore/issues/59443)

> ## Configure custom security descriptors for HTTP.sys request queues

Reviewer: Get rid of the "Why this matters" and "Hot to use" headings and connect with prose instead.

- **`aspnetcore.components.circuit.duration`** - Measures circuit lifetime duration and provides total count

### Blazor tracing

The new activity tracing capabilities use the `Microsoft.AspNetCore.Components` activity source and provide three main types of activities:

**Circuit lifecycle tracing:**
- **`Microsoft.AspNetCore.Components.CircuitStart`** - Traces circuit initialization with format `Circuit {circuitId}`
  - Tags: `aspnetcore.components.circuit.id`
  - Links: HTTP activity

**Navigation tracing:**
- **`Microsoft.AspNetCore.Components.RouteChange`** - Tracks route changes with format `Route {route} -> {componentType}`
  - Tags: `aspnetcore.components.circuit.id`, `aspnetcore.components.type`, `aspnetcore.components.route`
  - Links: HTTP trace, circuit trace

**Event handling tracing:**
- **`Microsoft.AspNetCore.Components.HandleEvent`** - Traces event handling with format `Event {attributeName} -> {componentType}.{methodName}`
  - Tags: `aspnetcore.components.circuit.id`, `aspnetcore.components.type`, `aspnetcore.components.method`, `aspnetcore.components.attribute.name`, `error.type`
  - Links: HTTP trace, circuit trace, router trace

> To enable Blazor metrics and tracing in your application, configure OpenTelemetry in your `Program.cs`:

Reviewer: ```suggestion
To enable Blazor metrics and tracing in your application, configure OpenTelemetry with the following meters and activity sources:
```

# ASP.NET Core in .NET 10 Preview 5 - Release Notes

Here's a summary of what's new in ASP.NET Core in this preview release:

- [Configure custom security descriptors for HTTP.sys request queues](#configure-custom-security-descriptors-for-httpsys-request-queues)
- [Validation resolver APIs marked as experimental](#validation-resolver-apis-marked-as-experimental)
- [Support for generating OpenAPI 3.1](#support-for-generating-openapi-31)
- [OpenAPI metadata from XML doc comments](#openapi-metadata-from-xml-doc-comments)
- [Add a Not Found page using the Blazor Router](#add-a-not-found-page-using-the-blazor-router)
- [Blazor metrics and tracing](#blazor-metrics-and-tracing)

ASP.NET Core updates in .NET 10:

- [What's new in ASP.NET Core in .NET 10](https://learn.microsoft.com/aspnet/core/release-notes/aspnetcore-10.0) documentation.
- [Breaking changes](https://docs.microsoft.com/dotnet/core/compatibility/10.0#aspnet-core)
- [Roadmap](https://github.com/dotnet/aspnetcore/issues/59443)

## Configure custom security descriptors for HTTP.sys request queues

> You can now specify a custom security descriptor for HTTP.sys request queues using the new `RequestQueueSecurityDescriptor` property on `HttpSysOptions`. This feature enables more granular control over access rights for the request queue, helping you tailor security to your application's needs.

Reviewer: ```suggestion
You can now specify a custom security descriptor for HTTP.sys request queues using the new `RequestQueueSecurityDescriptor` property on `HttpSysOptions`. This feature enables more granular control over access rights for the request queue, allowing you to tailor security to your application's needs.
```

// Create a discretionary access control list (DACL)
var dacl = new DiscretionaryAcl(isContainer: false, isDS: false, capacity: 2);
dacl.AddAccess(
    AccessControlType.Allow,
    new SecurityIdentifier(WellKnownSidType.BuiltinUsersSid, null),
    -1,
    InheritanceFlags.None,
    PropagationFlags.None
);
dacl.AddAccess(
    AccessControlType.Deny,
    new SecurityIdentifier(WellKnownSidType.BuiltinGuestsSid, null),
    -1,
    InheritanceFlags.None,
    PropagationFlags.None
);

// Assign the DACL to the security descriptor
securityDescriptor.DiscretionaryAcl = dacl;

// Configure HTTP.sys options
var builder = WebApplication.CreateBuilder();
builder.WebHost.UseHttpSys(options =>
{
    options.RequestQueueSecurityDescriptor = securityDescriptor;
});
```

### Additional notes

- The `RequestQueueSecurityDescriptor` applies only when creating a new request queue.
- This property does not affect existing request queues.
- See the official documentation for more information about Windows security descriptors and their usage.

## Validation resolver APIs marked as experimental

> To support future work on the new validation APIs in .NET 10, the underlying validation resolver APIs used to support minimal API validation have been marked as experimental. However, the top-level `AddValidation` APIs and the built-in validation filter for Minimal APIs are still non-experimental.

Reviewer 2: > To support future work on the new validation APIs in .NET 10 ... have been marked as experimental

This sentence is a bit off? Isn't it more "To support future work on the APIs in .NET 11 ..." if we want to stay vague, or "With some planned breaking changes in .NET 11 we're marking these APIs experimental in .NET 10 to allow users to make use of the feature but to signify that we're likely going to change the APIs in the future ..."

Reviewer: ```suggestion
To allow for future adjustments to the new validation APIs, the underlying validation resolver APIs used to support minimal API validation have been marked as experimental. However, the top-level `AddValidation` APIs and the built-in validation filter for Minimal APIs are still non-experimental.
```

> To allow for future adjustments to the new validation APIs, the underlying validation resolver APIs used to support minimal API validation have been marked as experimental. However, the top-level `AddValidation` APIs and the built-in validation filter for Minimal APIs are still non-experimental.

Reviewer: ```suggestion
To allow for future adjustments to the new validation APIs, the underlying validation resolver APIs used to support minimal API validation have been marked as experimental. However, the top-level `AddValidation` APIs and the built-in validation filter for Minimal APIs remain non-experimental.
```

## Support for generating OpenAPI 3.1

The OpenAPI.NET library used in ASP.NET Core OpenAPI document generation has been upgraded to [v2.0.0-preview18](https://github.com/microsoft/OpenAPI.NET/releases/tag/v2.0.0-preview.18).

## OpenAPI metadata from XML doc comments

> Support for generating OpenAPI metadata from XML doc comments has been extended to extract metadata for operation responses from `<returns>` and `<response>` XML tags.

Reviewer 3: ```suggestion
Support for generating OpenAPI metadata from XML doc comments has been extended to extract metadata for operation responses from `<returns>` and `<response>` XML tags on handler methods.
```

```html
<Router AppAssembly="@typeof(Program).Assembly" NotFoundPage="typeof(Pages.NotFound)">
    <Found Context="routeData">
        <RouteView RouteData="@routeData" />
        <FocusOnNavigate RouteData="@routeData" Selector="h1" />
    </Found>
    <NotFound>This content will be ignored because we have NotFoundPage defined.</NotFound>
</Router>
```

The Blazor project template now includes a `NotFound.razor` page by default. This page will automatically render whenever `NavigationManager.NotFound()` is called in your application, making it easier to handle missing routes with a consistent user experience.

## Blazor metrics and tracing

.NET 10 Preview 5 introduces comprehensive metrics and tracing capabilities for Blazor applications, providing detailed observability into component lifecycle, navigation, event handling, and circuit management.

### Blazor metrics

> The new metrics feature includes several meters to track different aspects of Blazor application performance:

Reviewer: ```suggestion
The new metrics feature includes several meters that track different aspects of Blazor application performance:
```

```csharp
builder.Services.ConfigureOpenTelemetryMeterProvider(meterProvider =>
{
    meterProvider.AddMeter("Microsoft.AspNetCore.Components");
    meterProvider.AddMeter("Microsoft.AspNetCore.Components.Lifecycle");
    meterProvider.AddMeter("Microsoft.AspNetCore.Components.Server.Circuits");
});

builder.Services.ConfigureOpenTelemetryTracerProvider(tracerProvider =>
{
    tracerProvider.AddSource("Microsoft.AspNetCore.Components");
});
```

These new observability features help you monitor and diagnose Blazor application performance, track user interactions, and understand component behavior in production environments.

## Contributors

Thank you contributors! ❤️

- [@am11](https://github.com/dotnet/aspnetcore/pulls?q=is%3Apr+is%3Amerged+milestone%3A10.0-preview5+author%3Aam11)
- [@Dona278](https://github.com/dotnet/aspnetcore/pulls?q=is%3Apr+is%3Amerged+milestone%3A10.0-preview5+author%3ADona278)
- [@feiyun0112](https://github.com/dotnet/aspnetcore/pulls?q=is%3Apr+is%3Amerged+milestone%3A10.0-preview5+author%3Afeiyun0112)
- [@MohabASHRAF-byte](https://github.com/dotnet/aspnetcore/pulls?q=is%3Apr+is%3Amerged+milestone%3A10.0-preview5+author%3AMohabASHRAF-byte)

> - [@profet23](https://github.com/dotnet/aspnetcore/pulls?q=is%3Apr+is%3Amerged+milestone%3A10.0-preview5+author%3Apropet23)

Reviewer: ```suggestion
- [@profet23](https://github.com/dotnet/aspnetcore/pulls?q=is%3Apr+is%3Amerged+milestone%3A10.0-preview5+author%3Aprofet23)
```

## OpenAPI metadata from XML doc comments

Support for generating OpenAPI metadata from XML doc comments has been extended to extract metadata for operation responses from `<returns>` and `<response>` XML tags.

## Add a Not Found page using the Blazor Router

Rendering content after triggering `NavigationManager.NotFound()` method can be now handled by passing a parameter with page type to the `Router`. It is a preferred way over `NotFound` fragment because it supports routing that can be used across different applications code re-execution middleware, including non-blazor ones. If `NotFound` fragment is defined together with `NotFoundPage`, the page has higher priority.

```html
<Router AppAssembly="@typeof(Program).Assembly" NotFoundPage="typeof(Pages.NotFound)">
    <Found Context="routeData">
        <RouteView RouteData="@routeData" />
        <FocusOnNavigate RouteData="@routeData" Selector="h1" />
    </Found>
    <NotFound>This content will be ignored because we have NotFoundPage defined.</NotFound>
</Router>
```

> ## Blazor metrics and tracing

Reviewer: There are too many subheadings in this section. Subheadings in release notes shouldn't be deeper than H3. I recommend having at most two subheadings in this section: one for the new metrics and one for the new tracing.
